### PR TITLE
feat: use zod to validate query output and infer query types (fixes #364)

### LINKED ISSUE #364: integrate zod into Slonik
At the moment, we can provide type suggestions the following way:

```ts
const persons = await pool.any<{id: number, name: string}>(sql`
  SELECT id, name
  FROM person
`);
```

While this works, there are no guarantees that this is what the query actually returns.

In order to add more excellent type safety, including runtime safety, Slonik is going to adopt [zod](https://www.npmjs.com/package/zod) natively.

The current thinking is that the API is going to be:

```ts
const persons = await pool.any(sql`
  ${z.object({
    id: z.number(),
    name: z.string(),
  })}
  SELECT id, name
  FROM person
`);
```

The reason for adding `z.object` to the `sql` template itself is because, ideally, we want them to live together.

We can then toggle runtime validation to ensure that queries return exactly what we type them to return, i.e. If this query were to change to:

```ts
const persons = await pool.any(sql`
  ${z.object({
    id: z.number(),
    name: z.string(),
  })}
  SELECT id, name, description
  FROM person
`);
```

We would throw an error about the unknown `description` property.

In contrast to build-time type safety, this mechanism of validation ensures that the application continues to work as expected (or breaks loudly) if database schema drifts from whatever we typed at build time, e.g. In the earlier example, we may have assumed that `name` is non-nullable based on our current knowledge of schema, but if that changed during the lifetime of the application, we'd want to know about it.

### PR DESCRIPTION
**Pull request by gajus**

### CODE REVIEW

File: src/factories/createSqlTag.ts [hunk 1]
```diff
@@ -47,8 +50,20 @@ const log = Logger.child({
 
 const sql: SqlTaggedTemplate = (
   parts: readonly string[],
-  ...values: readonly ValueExpression[]
+  ...args: readonly [ValueExpression | ZodTypeAny, ...ValueExpression[]]
```
Comment (mmkal) [inline src/factories/createSqlTag.ts:53]: Question re this api design. It's a little confusing having a zod object seemingly randomly in there with the query, and opens the door to incorrect stuff like

```ts
sql`
select ${z.object({ foo: z.string() })}, 123
```

could a similar goal be achieved with something like

```ts
const getSql = <T>(type: z.ZodType<T>) => {
  return (...args: ValueExpression[]) => ({
    ...sql<T>(...args),
    zodObject: type,
  })
}
```

then if you do something like `sql.type = getSql` you could use like:

```ts
const foo = sql.type(z.object({ foo: z.string() }))`
  select 'abc' as foo
`
```

Or if the type is big:

```ts
const Foo = z.object({ foo: z.string() })

const foo = sql.type(Foo)`
  select 'abc' as foo
`
```

that would also allow for other validation libraries that confirmed to Zod's interface. Wouldn't need to do instanceof type checks.

Comment (gajus) [reply src/factories/createSqlTag.ts:53]: This makes sense.

Comment (gajus) [reply src/factories/createSqlTag.ts:53]: After a bit of testing, going this direction. Thanks @mmkal !

**Comment by gajus** [comment]
Will start a new PR.

**Pull request closed without merging**